consensus/spos/worker.go

```diff
@@ -711,6 +728,47 @@ func (wrk *Worker) ResetConsensusMessages() {
 	wrk.consensusMessageValidator.resetConsensusMessages()
 }
 
+func (wrk *Worker) processEquivalentMessage(msgType consensus.MessageType, blockHeaderHash []byte) error {
+	// early exit if the message is not with final info
+	if !wrk.consensusService.IsMessageWithFinalInfo(msgType) {
+		return nil
+	}
+
+	hdrHash := string(blockHeaderHash)
+
+	wrk.mutEquivalentMessages.Lock()
+	defer wrk.mutEquivalentMessages.Unlock()
+
+	// if an equivalent message was seen before, return error to stop further broadcasts
+	numMessages := wrk.equivalentMessages[hdrHash]
```

Comment:
Is the validation of the equivalent message in a different PR?
Here if for example the first message received is invalid, you still count it.

Reply:
right, handled invalid message received in case of validation error